homeassistant/components/switch/rainmachine.py

```diff
@@ -1,208 +1,315 @@
+            dispatcher_send(self.hass, PROGRAM_UPDATE_TOPIC)
+        except HTTPError as exc_info:
             _LOGGER.error('Unable to turn on program "%s"', self.unique_id)
             _LOGGER.debug(exc_info)
 
-    @Throttle(MIN_SCAN_TIME, MIN_SCAN_TIME_FORCED)
     def update(self) -> None:
         """Update info for the program."""
-        import regenmaschine.exceptions as exceptions
+        from regenmaschine.exceptions import HTTPError
 
         try:
-            self._entity_json = self._client.programs.get(
-                self.rainmachine_entity_id)
-        except exceptions.HTTPError as exc_info:
+            self._obj = self.rainmachine.client.programs.get(
+                self._rainmachine_entity_id)
+
+            self._attrs.update({
+                ATTR_ID: self._obj['uid'],
+                ATTR_CS_ON: self._obj.get('cs_on'),
+                ATTR_CYCLES: self._obj.get('cycles'),
+                ATTR_DELAY: self._obj.get('delay'),
+                ATTR_DELAY_ON: self._obj.get('delay_on'),
+                ATTR_SOAK: self._obj.get('soak'),
+                ATTR_STATUS:
+                    PROGRAM_STATUS_MAP[self._obj.get('status')],
+                ATTR_ZONES: ', '.join(z['name'] for z in self.zones)
+            })
+        except HTTPError as exc_info:
             _LOGGER.error('Unable to update info for program "%s"',
                           self.unique_id)
             _LOGGER.debug(exc_info)
 
 
-class RainMachineZone(RainMachineEntity):
+class RainMachineZone(RainMachineSwitch):
     """A RainMachine zone."""
 
-    def __init__(self, client, device_mac, zone_json,
-                 zone_run_time):
+    def __init__(self, rainmachine, obj, zone_run_time):
         """Initialize a RainMachine zone."""
-        super().__init__(client, device_mac, zone_json)
+        super().__init__(rainmachine, 'zone', obj)
+
+        self._properties_json = {}
         self._run_time = zone_run_time
-        self._attrs.update({
-            ATTR_CYCLES: self._entity_json.get('noOfCycles'),
-            ATTR_TOTAL_DURATION: self._entity_json.get('userDuration')
-        })
 
     @property
     def is_on(self) -> bool:
         """Return whether the zone is running."""
-        return bool(self._entity_json.get('state'))
+        return bool(self._obj.get('state'))
 
     @property
     def name(self) -> str:
         """Return the name of the zone."""
-        return 'Zone: {0}'.format(self._entity_json.get('name'))
+        return 'Zone: {0}'.format(self._obj.get('name'))
 
-    @property
-    def unique_id(self) -> str:
-        """Return a unique, HASS-friendly identifier for this entity."""
-        return '{0}_zone_{1}'.format(
-            self.device_mac.replace(':', ''), self.rainmachine_entity_id)
+    @callback
+    def _program_updated(self):
+        """Update state, trigger updates."""
+        self.schedule_update_ha_state(True)
```

Comment:
Use `async_schedule_update_ha_state`.

Reply:
Got it. For my own knowledge, this is because the `@callback` decorator declares that anything in the method body is safe for the event loop, yes?

Reply:
Yes! When the dispatcher calls this method it uses the core method `async_add_job` which checks if the function/method has been decorated with `callback`, and if so, schedules a call to it on the event loop. Otherwise it will be run in the thread pool (if it's not a coroutine or coroutine function).